nomenclate/core/nomenclature.py

```diff
@@ -2,146 +2,295 @@
+            else:
+                raise exceptions.ValidationError('Invalid type %s, expected %s' % (type(entry), str))
+
+    def __eq__(self, other):
+        return self.token == other.token and self.label == other.label
+
+    def __repr__(self):
+        return '%s:%s' % (self.token, self.label)
+
+
+class TokenAttrDict(dict):
+    def __init__(self, nomenclate_object, signal):
+        super(dict, self).__init__()
+        self.nom = nomenclate_object
+        self.signal = signal
+
+    @property
+    def state(self):
+        return dict((name_attr.token, name_attr.label) for name_attr in self.get_token_attrs())
+
+    @state.setter
+    def state(self, input_object, **kwargs):
+        """
+        Args:
+            input_object Union[dict, self.__class__]: accepts a dictionary or self.__class__
+        """
+        if not isinstance(input_object, dict):
+            raise exceptions.FormatError('state setting only accepts type %s, not given input %s %s' %
+                                         (dict, input_object, type(input_object)))
+
+        if isinstance(input_object, self.nom.__class__):
+            input_object = input_object.state
+
+        input_object.update(kwargs)
+
+        for input_attr_name, input_attr_value in iteritems(input_object):
+            self.set_token_attr(input_attr_name, input_attr_value)
+
+    def build_name_attrs(self):
+        """ Creates all necessary name attributes for the naming convention
+        """
+        self.purge_invalid_name_attrs()
+
+        for token in self.nom.format_order:
+            self.set_token_attr(token, "")
+
+    def purge_invalid_name_attrs(self):
+        """ Removes name attrs not found in the format order
+        """
+        for token_attr in self.get_token_attrs():
+            try:
+                self._validate_name_in_format_order(token_attr.label, self.nom.format_order)
+            except exceptions.FormatError:
+                del self[token_attr.token]
+
+    def purge_name_attrs(self):
+        for token_attr in self.get_token_attrs():
+            del self[token_attr.token]
+        self.signal.fire()
+
+    def clear_name_attrs(self):
+        for token_attr in self.get_token_attrs():
+            token_attr.set('')
+
+    def set_token_attr(self, token, value):
+        token_attrs = self.get_token_attrs()
+        if token not in [token_attr.token for token_attr in token_attrs]:
+            self._create_token_attr(token, value)
+            self.signal.fire()
+        else:
+            for token_attr in token_attrs:
+                if token == token_attr.token:
+                    setattr(self, token_attr.token, value)
+
+    def get_token_attr(self, token):
+        token_attr = getattr(self, token.lower())
+        if token_attr is None:
+            raise exceptions.SourceError('This nomenclate instance has no %s attribute set.' % token)
+        else:
+            return token_attr
+
+    def get_token_attrs(self):
```

Comment:
No, returning the attribute object, it's confusing since self is a dict so the actual key is just the string attr representation.